analysis/semsim/utils.py

```diff
@@ -0,0 +1,59 @@
+import gcld3
+import pandas as pd
+from sklearn.model_selection import train_test_split
+from analysis.semsim import unifiedmodel
+from torch.utils.data import DataLoader
+import pytorch_lightning as pl
+import torch
+from sklearn.metrics import roc_auc_score
+import numpy as np
+from scipy.special import expit
+
+# Remove all pairs from data that have languages other than that specified.
+def filter_lang(data, with_transcript, lang):
+    detector = gcld3.NNetLanguageIdentifier(min_num_bytes=0, max_num_bytes=1000)
+    processed_data = data.copy(deep=True)
+    for part in ['regret_title', 'recommendation_title', 'regret_description', 'recommendation_description'] + (['regret_transcript', 'recommendation_transcript'] if with_transcript else []):
+        processed_data = processed_data[
+            (processed_data[part].str.len() < 10) |
+            [detector.FindLanguage(text=x).language == lang for x in processed_data[part]]
+        ]
+    return processed_data
+
+def run_experiment(data, with_transcript, sample_negative, epochs, batch_size, lr, freeze_policy):
+    exp_data = data.copy(deep=True)
+    if not with_transcript:
+        exp_data = exp_data[exp_data.regret_transcript.isnull() | exp_data.recommendation_transcript.isnull()]
+        exp_data.drop(['regret_transcript', 'recommendation_transcript'], axis=1, inplace=True)
+
+    exp_data = exp_data.dropna()
+
+    if sample_negative:
+        exp_data = pd.concat([exp_data.query("label==1"), exp_data.query("label==0").sample(len(exp_data.query("label==1")))])
+    rest_data, test_data = train_test_split(exp_data, test_size=0.25, stratify=exp_data.label)
+    train_data, val_data = train_test_split(rest_data, test_size=0.33, stratify=rest_data.label)
+    del exp_data
+
+    train_dataset = unifiedmodel.RRUMDatasetArrow(train_data, with_transcript=with_transcript)
+    val_dataset = unifiedmodel.RRUMDatasetArrow(val_data, with_transcript=with_transcript)
+
+    train_loader = DataLoader(train_dataset.encoded_dataset, shuffle=True, batch_size=batch_size, num_workers=0, pin_memory=False)
+    val_loader = DataLoader(val_dataset.encoded_dataset, shuffle=False, batch_size=batch_size, num_workers=0, pin_memory=False)
+    
+    model = unifiedmodel.RRUM(
+        with_transcript=with_transcript,
+        optimizer_config=(lambda x: torch.optim.Adam(x.parameters(), lr=lr)),
```

Comment:
If we change the optimizer config as in earlier comments, this could be changed to
```
optimizer=optimizer,
lr=lr
```
and add optimizer to `run_experiment()` arguments. The value of optimizer could then be changed when running different experiments